### PR TITLE
turn off sqlite auto_checkpoint and introduce period-based checkpoint calls

### PR DESCRIPTION
**Pull request by Horusiath**
This PR turns off auto_checkpoint in sqld-embed SQLite instance (which by default is every 1000 pages) and introduces a fiber loop, which checkpoints given configured time interval (by default it's every 1h).

### CODE REVIEW

File: sqld/src/lib.rs [hunk 1]
```diff
@@ -587,6 +591,54 @@ async fn run_storage_monitor(db_path: PathBuf, stats: Stats) -> anyhow::Result<(
     Ok(())
 }
 
+async fn run_checkpoint_cron(db_path: PathBuf, period: Duration) -> anyhow::Result<()> {
+    const RETRY_INTERVAL: Duration = Duration::from_secs(60);
+    let data_path = db_path.join("data");
+    tracing::info!("setting checkpoint interval to {:?}", period);
+    let mut interval = interval(period);
```
Comment (psarna) [inline sqld/src/lib.rs:598]: The default "missed tick behavior" for Tokio interval is to aggresively catch up, ticking multiple times: https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html . That makes little sense with checkpointing, and it can be a performance issue, since we'd repeatedly take a write lock multiple times in a row. Judging from the docs, perhaps https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Delay is a better choice?

File: sqld/src/replication/primary/logger.rs [hunk 2]
```diff
@@ -897,6 +897,14 @@ fn checkpoint_db(data_path: &Path) -> anyhow::Result<()> {
             );
             Ok(())
         })?;
+        // turn off auto_checkpointing - we'll use a fiber to checkpoint in time steps instead
+        let rc = rusqlite::ffi::sqlite3_wal_autocheckpoint(conn.handle(), 0);
+        if rc != 0 {
+            return Err(anyhow!(
```
Comment (psarna) [inline sqld/src/replication/primary/logger.rs:903]: nitpick/hint: this is equivalent to shorter: `bail!("Failed to disable WAL autocheckpoint - error code: {rc}")`. Also applies to other conditions below

File: sqld/src/replication/primary/logger.rs [hunk 3]
```diff
@@ -897,6 +897,14 @@ fn checkpoint_db(data_path: &Path) -> anyhow::Result<()> {
             );
             Ok(())
         })?;
+        // turn off auto_checkpointing - we'll use a fiber to checkpoint in time steps instead
+        let rc = rusqlite::ffi::sqlite3_wal_autocheckpoint(conn.handle(), 0);
```
Comment (psarna) [inline sqld/src/replication/primary/logger.rs:901]: I saw later that the autocheckpoint interval is also allowed to be off (it's an Option<u64>). In that case, I suppose we should *not* turn off autocheckpointing, because it would make the database never checkpoint and just go out of storage. And if somebody wants that behavior, they can set the checkpoint interval to 1 billion hours.

**Comment by psarna** [comment]
@MarinPostma this commit makes sqld take over checkpointing, which is what we discussed a few times in context of deduplicating WAL. I'd appreciate you take a look to see if it doesn't interfere with replication in any way (it shouldn't, but worth double-checking)

**Review by psarna — DISMISSED** [review]
@Horusiath also, we should turn off auto_checkpoint in every single database connection that opens, because by default, each connection opens with autocheckpoint set to 1000 pages. So, all places that call libsql_open, like https://github.com/libsql/sqld/blob/f50803f71f81bd0333972301b0c4f9de66026ccb/sqld-libsql-bindings/src/lib.rs#L90, need to run `sqlite3_wal_autocheckpoint(db, 0)` as well. If we miss even one, it will cause a checkpoint to happen eventually

**Comment by MarinPostma** [comment]
we should limit the WAL size as well. The disk size on the platform is rather, and the WAL can grow very fast

File: sqld/src/lib.rs [hunk 4]
```diff
@@ -587,6 +591,54 @@ async fn run_storage_monitor(db_path: PathBuf, stats: Stats) -> anyhow::Result<(
...
+        if let Some(retry) = retry.take() {
+            sleep(retry).await;
+        } else {
+            interval.tick().await;
+        }
+        retry = match rusqlite::Connection::open(&data_path) {
+            Ok(conn) => unsafe {
+                let start = Instant::now();
+                let mut num_checkpointed: std::ffi::c_int = 0;
+                let rc = rusqlite::ffi::sqlite3_wal_checkpoint_v2(
```
Comment (MarinPostma) [inline sqld/src/lib.rs:610]: this is blocking

File: sqld/src/lib.rs [hunk 5]
```diff
@@ -587,6 +591,54 @@ async fn run_storage_monitor(db_path: PathBuf, stats: Stats) -> anyhow::Result<(
...
+    tracing::info!("setting checkpoint interval to {:?}", period);
+    let mut interval = interval(period);
+    let mut retry = None;
+    loop {
+        if let Some(retry) = retry.take() {
+            sleep(retry).await;
+        } else {
+            interval.tick().await;
+        }
+        retry = match rusqlite::Connection::open(&data_path) {
```
Comment (MarinPostma) [inline sqld/src/lib.rs:606]: this is blocking as well

File: sqld/src/replication/primary/logger.rs (hunk 2 above)
Comment (MarinPostma) [reply sqld/src/replication/primary/logger.rs:903]: yeah, and clippy should complain

File: sqld/src/replication/primary/logger.rs [hunk 6]
```diff
@@ -905,17 +913,22 @@ fn checkpoint_db(data_path: &Path) -> anyhow::Result<()> {
...
-        );
-
-        conn.execute("VACUUM", ())?;
+        if rc == 0 {
+            if num_checkpointed == -1 {
+                Err(anyhow!(
+                    "Checkpoint failed: database journal_mode is not WAL"
+                ))
+            } else {
+                conn.execute("VACUUM", ())?;
```
Comment (MarinPostma) [inline sqld/src/replication/primary/logger.rs:922]: not sure a vacuum is necessary on every checkpoint

File: sqld/src/replication/primary/logger.rs (hunk 2 above)
Comment (Horusiath) [reply sqld/src/replication/primary/logger.rs]: clippy doesn't complain. I've changed it anyway.

File: sqld/src/lib.rs (hunk 4 above)
Comment (Horusiath) [reply sqld/src/lib.rs:619]: I've wrapped it with spawn_blocking.

File: sqld/src/replication/primary/logger.rs (hunk 6 above)
Comment (Horusiath) [reply sqld/src/replication/primary/logger.rs:925]: That's the original code. I'm not sure, what would be the best course of action here. We can leave VACUUM as an explicit request from the user side.

Comment (MarinPostma) [reply sqld/src/replication/primary/logger.rs:925]: The intent was different, it was meant for replication log recovery

Comment (Horusiath) [reply sqld/src/replication/primary/logger.rs:925]: Maybe I'll just remove it and let users perform it explicitly. It shouldn't be a problem most of the time.

File: sqld/src/lib.rs (hunk 4 above)
Comment (psarna) [reply sqld/src/lib.rs:620]: @MarinPostma just to double-check -> this isn't running in Tokio main thread, so isn't `block_in_place` more lightweight?

**Comment by psarna** [comment]
> @Horusiath also, we should turn off auto_checkpoint in every single database connection that opens, because by default, each connection opens with autocheckpoint set to 1000 pages. So, all places that call libsql_open, like
> 
> https://github.com/libsql/sqld/blob/f50803f71f81bd0333972301b0c4f9de66026ccb/sqld-libsql-bindings/src/lib.rs#L90
> 
> , need to run `sqlite3_wal_autocheckpoint(db, 0)` as well. If we miss even one, it will cause a checkpoint to happen eventually

@Horusiath was the comment about optionally turning off autocheckpoint for every single connection that we open already applied, or not yet?

**Comment by Horusiath** [comment]
@psarna question would be - should we turn off auto_checkpoint by default on connection open? In general this only happens if checkpoint interval and bottomless are set up. Given that these can be set either as env vars or command line params, it gets extremely tedious to pass that information down to connection open.

**Comment by psarna** [comment]
@Horusiath to the best of my knowledge we open db connections in two main places for user traffic:
```
[sarna@sarna-pc sqld]$ git grep sqld_libsql_bindings::Connection::open\(
sqld/src/database/libsql.rs:    sqld_libsql_bindings::Connection::open(path, flags, wal_methods, hook_ctx)
sqld/src/replication/replica/injector.rs:        let conn = sqld_libsql_bindings::Connection::open(
```
and two places for our own maintenance ops:
```
[sarna@sarna-pc sqld]$ git grep rusqlite::Connection::open\(
sqld/src/main.rs:    let conn = rusqlite::Connection::open(db_path.join("data"))?;
sqld/src/replication/primary/logger.rs:        let conn = rusqlite::Connection::open(data_path)?;
```
, and moreover, I think we only ever intend to replicate to S3 from the primary, so the connection open in `replica/injector.rs` can be ignored for now.

That leaves 3 places to handle, and I think all of them at some point have access to our `Config` instance, which contains all the information we need to determine if we should disable autocheckpoint or not.

Since autocheckpoint is a per-connection option, and we want to control how often checkpoints happen, I don't see any other option rather than consistently disable it if needed for each connection that we open. There's also an SQLite compilation flag that allows changing the default threshold, but it's not that interesting from our point of view, since we need to be able to determine the autocheckpoint threshold from command-line parameters. Another option to consider would be to contribute to libSQL and add a global option that sets the default WAL autocheckpoint for new connections, but that sounds a little too specific to be generally useful, so I'm not sure. In any case, we need to be able to consistently disable autocheckpointing for all connections that we create, otherwise we'd have spurious checkpoints that we don't want. /cc @haaawk

**Comment by Horusiath** [comment]
Generally, `sqld_libsql_bindings::Connection::open` is used by:
1. `replica::injector::FrameInjector` - leave auto_checkpoint to default?
2. As part of `libsql::open_db` which is used in 5 different places:
 - `storage_monitor` which is read only (so we can leave autocheckpoint untouched)
 - `DumpLoader` which can execute any arbitrary scripts.
 - `libsql::Connection<'a>` which can be called from anywhere and doesn't seem to have access to bottomless config

**Comment by psarna** [comment]
@Horusiath after giving it some more thought, I think the following makes sense:
 > replica::injector::FrameInjector - leave auto_checkpoint to default?

yes, replicas don't use bottomless anyway

 > As part of libsql::open_db which is used in few different places:
 > storage_monitor which is read only (so we can leave autocheckpoint untouched)

yes

 > DumpLoader which can execute any arbitrary scripts.

does it only ever get called on startup, before bottomless is initialized? If so, we can leave autocheckpoint on, or always turn it off, I don't care much either way

 > libsql::Connection<'a> which can be called from anywhere and doesn't seem to have access to bottomless config
 
here's the only place where we actually need to push the configuration option, so that we know if we should enable or disable autocheckpoint, and to which value

... and let me also reiterate my mantra that the most important bit is to make sure that the checkpointing fiber uses a connection with our WAL methods from primary/logger.rs registered, so that it calls bottomless code on checkpoint.

**Comment by MarinPostma** [comment]
> does it only ever get called on startup, before bottomless is initialized? If so, we can leave autocheckpoint on, or always turn it off, I don't care much either way

actually, we probably shouldn't even for that, maybe a handle to manually trigger checkpoints that we are sure bottomless will pick up

**Comment by Horusiath** [comment]
@psarna could you review again?

File: sqld/src/main.rs [hunk 7]
```diff
@@ -182,6 +182,15 @@ struct Cli {
     #[clap(long, env = "SQLD_SNAPSHOT_EXEC")]
     snapshot_exec: Option<String>,
 
+    /// The address and port for the replication HTTP API.
+    #[clap(long, env = "SQLD_HTTP_REPLICATION_LISTEN_ADDR")]
+    http_replication_listen_addr: Option<SocketAddr>,
```
Comment (haaawk) [inline sqld/src/main.rs:187]: What is this?

File: sqld/src/main.rs [hunk 8]
```diff
@@ -301,6 +310,8 @@ fn config_from_args(args: Cli) -> Result<Config> {
         max_response_size: args.max_response_size.0,
         max_total_response_size: args.max_total_response_size.0,
         snapshot_exec: args.snapshot_exec,
+        http_replication_addr: args.http_replication_listen_addr,
```
Comment (haaawk) [inline sqld/src/main.rs:313]: What is this?

File: sqld/src/lib.rs [hunk 9]
```diff
@@ -105,6 +107,8 @@ pub struct Config {
     pub max_response_size: u64,
     pub max_total_response_size: u64,
     pub snapshot_exec: Option<String>,
+    pub http_replication_addr: Option<SocketAddr>,
```
Comment (haaawk) [inline sqld/src/lib.rs:110]: What is this?

File: sqld/src/lib.rs [hunk 10]
```diff
@@ -146,6 +150,8 @@ impl Default for Config {
             max_response_size: 10 * 1024 * 1024,       // 10MiB
             max_total_response_size: 32 * 1024 * 1024, // 32MiB
             snapshot_exec: None,
+            http_replication_addr: None,
```
Comment (haaawk) [inline sqld/src/lib.rs:153]: What is this?

File: sqld/src/lib.rs [hunk 11]
```diff
@@ -527,7 +534,7 @@ async fn run_storage_monitor(db_path: PathBuf, stats: Stats) -> anyhow::Result<(
             // initialize a connection here, and keep it alive for the entirety of the program. If we
             // fail to open it, we wait for `duration` and try again later.
             let ctx = &mut ();
-            let maybe_conn = match open_db(&db_path, &TRANSPARENT_METHODS, ctx, Some(rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)) {
+            let maybe_conn = match open_db(&db_path, &TRANSPARENT_METHODS, ctx, Some(rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY), DEFAULT_AUTO_CHECKPOINT) {
```
Comment (haaawk) [inline sqld/src/lib.rs:537]: Add a comment that this is read only so it won't generate any new frames and the auto_checkpoint setting can be whatever and we just pick DEFAULT_AUTO_CHECKPOINT but it does not matter as no checkpoint is ever performed when no writes appear on the connection.

File: sqld/src/replication/replica/injector.rs [hunk 12]
```diff
@@ -20,6 +21,7 @@ impl<'a> FrameInjector<'a> {
                 | OpenFlags::SQLITE_OPEN_NO_MUTEX,
             &INJECTOR_METHODS,
             hook_ctx,
+            DEFAULT_AUTO_CHECKPOINT,
```
Comment (haaawk) [inline sqld/src/replication/replica/injector.rs:24]: Please add a comment why it's ok to keep using default auto checkpoint for the replication. It's not obvious.

File: bottomless/src/replicator.rs [hunk 13]
```diff
@@ -77,6 +78,12 @@ pub struct Options {
     pub use_compression: CompressionKind,
     pub aws_endpoint: Option<String>,
     pub db_id: Option<String>,
+    /// AWS bottomless access key ID.
+    pub access_key_id: Option<String>,
+    /// AWS bottomless secret access key.
+    pub secret_access_key: Option<String>,
+    /// AWS bottomless region.
+    pub region: Option<String>,
```
Comment (haaawk) [inline bottomless/src/replicator.rs:86]: As discussed, probably better to do in a separate PR.

**Review by haaawk — CHANGES_REQUESTED** [review]
Looks mostly good except for one place when we're using QueryBuilderConfig::default in libsql.rs

File: bottomless/src/replicator.rs [hunk 14]
```diff
@@ -192,7 +192,7 @@ impl Default for Options {
         Options {
             create_bucket_if_not_exists: true,
             verify_crc: true,
-            use_compression: CompressionKind::None,
+            use_compression: CompressionKind::Gzip,
```
Comment (haaawk) [inline bottomless/src/replicator.rs:195]: How is this change related to auto checkponting?

Comment (haaawk) [inline bottomless/src/replicator.rs:195]: How is this change related to auto checkponiting?

**Review by haaawk — APPROVED** [review]

File: bottomless/src/replicator.rs (hunk 14 above)
Comment (Horusiath) [reply bottomless/src/replicator.rs:195]: I think, I've left it when started running comparisons with Litestream. Keeping non-compressed WAL frames around is extremely expensive and we definitely shouldn't do that by default.
